### PR TITLE
Use avvio lifecycle

### PR DESCRIPTION
**Pull request by delvedor**
With this pr we are introducing the [avvio](https://github.com/mcollina/avvio) lifecycle, in this way the internal code is more clean and it's easier handle the asynchronous bootstrap.

In addition, now UpRing exposes the `use` api, this means that we can build plugins!
A perfect example could be done with [upring-kw](https://github.com/upringjs/upring-kv) or [upring-pubsub](https://github.com/upringjs/upring-pubsub).

Feedbacks?

### CODE REVIEW

File: upring.js [hunk 1]
```diff
@@ -218,7 +140,7 @@ UpRing.prototype.mymeta = function () {
 }
 
 UpRing.prototype.request = function (obj, callback, _count) {
-  if (!this.ready) {
+  if (!this.isReady) {
```
Comment (mcollina) [inline upring.js]: can you restore this to `ready`? It seems unnecessary breakage.

File: lib/dispatch.js [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+        if (result && typeof result.then === 'function') {
+          result
+            .then(res => process.nextTick(reply, null, res))
+            .catch(err => process.nextTick(reply, err, null))
+        }
+      } else {
+        reply(new Error('message does not match any pattern'))
+      }
+    } else {
+      upring.emit('request', req, reply)
```
Comment (mcollina) [inline lib/dispatch.js]: can you use `this`  here, and not `upring`?

File: upring.js (hunk 1 above)
Comment (delvedor) [inline upring.js]: I can't because `avvio` exposes `ready` as a function :)

File: lib/dispatch.js [hunk 3]
```diff
@@ -0,0 +1,31 @@
+'use strict'
+
+module.exports = function dispatchPlugin (upring, opts, next) {
```
Comment (mcollina) [inline lib/dispatch.js]: actually, this is probably better attached to the main `upring` object.

**Comment by mcollina** [comment]
Can you add the docs in the README for avvio methods?

**Pull request merged**